### PR TITLE
Ekat library

### PR DESCRIPTION
**Pull request by jgfouca**
Split-out EKAT files into their own subdirectory as a preliminary step towards making it its own repo.

### CODE REVIEW

**Comment by jgfouca** [comment]
@jeff-cohere good question, "EKAT" hurts my eyes a bit.

**Comment by bartgol** [comment]
> Looks great. I might as well ask this now, before we proceed any further:
> 
> Given that we refer to SCREAM and Kokkos in the code as `scream` and `kokkos`, what do we think about using `ekat` instead of `EKAT`?
> 
> Also, just a question: are we going to rename the `scream_*` stuff to `ekat_*` once it's all separate?

I'm also for 'ekat' vs 'EKAT'. spares me an inconvenient shift key (long) press. :)

Also, I think we'll rename those files, yes. Not sure if it's best to do it in a second PR.

**Comment by jgfouca** [comment]
I expect some controversy over scream_session, scream_catch_main, and maybe scream_types and scream_assert to a lesser extent.

I think scream_session can be generally useful as encapsulating initializing a kokkos-based system. Similarly, scream_catch_main is useful for generic MPI+catch process. Obviously, the file names, function names, and namespace will have to be renamed.

**Comment by bartgol** [comment]
I started a review, then since I had many comments, I stopped, and decided to work a bit on this.

Bottom line is that ekat should not depend on scream_config.h stuff. That is scream specific. Some things could be ported into ekat, and have ekat produce a ekat_config.h file, but I'm not sure what belongs there. For instance, does the Real type definition belong to ekat? I tend to say no. Downstream packages can define that type. What about default FPEs? Ekat can handle turning on/off exceptions, but downstream projects should define what their default fpes are, not ekat.

In short, I am going to strip some stuff off of ekat, and put it back in scream. While at it, I may rename some ekat file from scream_xyz to ekat_xyz.

**Comment by jgfouca** [comment]
@bartgol , yes, I had the same thoughts regarding scream_config.hpp.

**Comment by jgfouca** [comment]
GitHub just threw away my edits, so I will try a new comment.

I think virtually all clients will want FPE and precision controls, so why not put that stuff in ekat? The client would control the setting via the cmake configure command.

**Comment by bartgol** [comment]
> GitHub just threw away my edits, so I will try a new comment.
> 
> I think virtually all clients will want FPE and precision controls, so why not put that stuff in ekat? The client would control the setting via the cmake configure command.

I think we can leave the 'generic' FPE functionalities there, but remove the 'default' fpe stuff. Each project/folder should decide what are their desired fpe (if any), and enable them. We may even want different fpes for different unit tests.

As for the precision, I think the tools are useful. What concerns me is that putting the Real type definition in ekat forces all users of ekat to use the same precision. I can foresee us experimenting a parametrization in single precision, and another in double. If we start using the 'Real' type across ekat, we force all users of ekat to use the same precision.

My preferred approach would be to template ekat's stuff over the real type (whenever needed), and let the users decide what precision (i.e., real type) to use.

**Comment by jeff-cohere** [comment]
I agree that we should give individual packages the ability to set their precision for floating point numbers. This is an item of discussion in EAGLES.

**Comment by bartgol** [comment]
The only real trouble are the f90 interoperable routines. F90 and C do not have overloads, so you can only have one routine called `void foo (sometype* data);`, and you need to decideif sometype is single or double.

When goint from f90 to C++, a solution is to use a `void*`, and pass also an integer, corresponding to the size (in byte) of the scalar type, then, in C, call a C++ templated function, with the proper template type based on the scalar size.

When going from C++ to f90, we can use a generic function, that then dispatches to `blah_sp` and `blah_dp` depending again on a scalar size input value, which can be either 4 (sp) or 8 (dp).

**Comment by bartgol** [comment]
After discussing with Jim, we decided we can keep precision in the ekat configuration. If another package needs ekat built with a different configuration, we can build it with a different precision. So one can do

```
set (EKAT_DOUBLE_PRECISION ON)
add_subdirectory (path/to/ekat ${CMAKE_BINARY_DIR}/ekat_dp)

set (EKAT_DOUBLE_PRECISION OFF)
add_subdirectory (path/to/ekat ${CMAKE_BINARY_DIR}/ekat_sp)
```
So long as one does not add both bin folders (where ekat_config.h resides) to its cmake include dirs, it should be fine.

**Comment by bartgol** [comment]
Ok, I did a big chunk of work to pacakge ekat as much as possible as an independent package. All files and namespaces still have to be adjusted, but that's just a matter of sed-ing.

I am off tomorrow, and Monday is holiday. I may or may not work a bit on this during the long weekend, but can't guarantee.

Tests in the ekat package are not yet passing. I think we need to set compiler flags _before_ processing ekat, or, better, _while_ processing it. But I need to make sure that after adding ekat subdir we don't perform logic that would influence the flags.

Right now, ekat takes care of including the tpls for which it offers utilities/support (kokkos and yamlcpp).
_However_, one can use ekat's cmake scripts to process kokkos beforehand. These scripts make sure that a src folder is not processed more than once, by using cmake global properties (better than cache variables in this context). So one _could_ process kokkos _before_ processing ekat. The latter would simply needs the Kokkos_blah variables set when it's time to add the ekat library, it doesn't matter if kokkos is processed by ekat or before.

**Comment by bartgol** [comment]
I think this PR is ready for integration. All tests pass on my machine in DP (I still have to check SP and FPE, but I might let the AT do that).

This PR is already quite big, but it's not the whole work. Missing items:

- rename files/functions/namespaces to purge the 'scream' acronym from Ekat
- package ekat in a submodule, to ship inside e3sm (rather than a scream subfolder).

Those items will be addressed in a following PR, to avoid this becoming too big.

**Comment by bartgol** [comment]
I saw the github warning that this branch has conflicts that need to be resolved. I tried `git merge origin/master` on my computer and it went through without causing a conflict. The AT should be able to test the branch, but we might have to manually merge into master, and then push (not sure if that will require disabling master protection, but I would like to avoid the extra commit coming from merging master in the branch and then branch in master).

**Comment by bartgol** [comment]
Note: I removed myself as reviewer, since I did a big chunk of the mods, and makes no sense that I review myself.

File: components/scream/src/share/grid/remap/abstract_remapper.hpp [hunk 1]
```diff
@@ -3,11 +3,11 @@
 
 #include "share/field/field.hpp"
 #include "share/field/field_utils.hpp"
-#include "share/util/scream_factory.hpp"
-#include "share/util/string_utils.hpp"
-#include "share/util/scream_std_utils.hpp"
-#include "share/scream_parameter_list.hpp"
 #include "share/grid/abstract_grid.hpp"
+#include "ekat/util/scream_factory.hpp"
```
Comment (jeff-cohere) [inline components/scream/src/share/grid/remap/abstract_remapper.hpp:7]: Scream factory sounds like a bad horror film. :-)

File: components/scream/src/physics/shoc/atmosphere_macrophysics.hpp [hunk 2]
```diff
@@ -11,11 +11,11 @@ namespace scream
...
 /*
  * The class responsible to handle the atmosphere microphysics
- * 
+ *
  * The AD should store exactly ONE instance of this class stored
  * in its list of subcomponents (the AD should make sure of this).
- * 
- *  Note: for now, scream is only going to accommodate SHOC and as macrophysics 
+ *
+ *  Note: for now, scream is only going to accommodate SHOC and as macrophysics
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/atmosphere_macrophysics.hpp:18]: Tiny nitpick: SHOC and what? Can we strike the "and" here?

File: components/scream/ekat/tests/algorithm/lin_interp_test.cpp [hunk 3]
```diff
@@ -176,11 +176,16 @@ TEST_CASE("lin_interp", "soak") {
 
     // Compare results
     {
+#ifdef EKAT_STRICT_FP
```
Comment (jeff-cohere) [inline components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: Is this macro defined only for this test, or as part of the EKAT configuration? If the latter, maybe we need some sort of value for Strict exposed that prevents folks from having to do this #ifdef themselves. Sorry if this is a silly question--there's a lot of code in this PR. :-)

File: components/scream/ekat/cmake/Ekat.cmake [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+#
+# However, before adding the subdir, you need to make sure that
+# the binary folder is 'available', that is, nobody has yet tried to
+# build ekat with that precision. This script can do that for you
+
+# Define global properties for EKAT_DOUBLE_BUILT and EKAT_SINGLE_BUILT
+# to ensure ekat (with the desired precision) is built only once
+define_property(GLOBAL
+                PROPERTY EKAT_DOUBLE_BUILT
+                BRIEF_DOCS "Wheter ekat subdir (with precision set to double) has already been processed"
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/Ekat.cmake:26]: Wheter -> Whether

File: components/scream/ekat/cmake/Ekat.cmake [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                FULL_DOCS "This property is used by cmake to ensure that EKAT
+                           submodule directory is only processed once (with add_subdirectory).")
+define_property(GLOBAL
+                PROPERTY EKAT_SINGLE_BUILT
+                BRIEF_DOCS "Wheter ekat subdir (with precision set to single) has already been processed"
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/Ekat.cmake:31]: Whether

File: components/scream/ekat/cmake/EkatCreateUnitTest.cmake [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        set_tests_properties(${FULL_TEST_NAME} PROPERTIES LABELS "${ecut_LABELS}")
+      endif()
+
+      math(EXPR CURR_THREADS "${CURR_THREADS}+${THREAD_INCREMENT}")
+    endwhile()
+    set(CURR_THREADS ${THREAD_START})
+    math(EXPR CURR_RANKS "${CURR_RANKS}+${MPI_INCREMENT}")
+  endwhile()
+
+endfunction(EkatCreateUnitTest)
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/EkatCreateUnitTest.cmake:202]: This function will be very useful!

File: components/scream/ekat/cmake/EkatMpiUtils.cmake [hunk 7]
```diff
@@ -1,28 +1,29 @@
...
+    set (${DISTRO_NAME} "openmpi")
   else ()
     execute_process(COMMAND ${CMAKE_CXX_COMPILER} 
                     ${CMAKE_CURRENT_SOURCE_DIR}/cmake/TryCompileMPICH.cpp
                     OUTPUT_QUIET ERROR_QUIET
                     RESULT_VARIABLE HAVE_MPICH_MPI)
 
     if (HAVE_MPICH_MPI)
-      set (DISTRO_NAME "mpich")
+      set (${DISTRO_NAME} "mpich")
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/EkatMpiUtils.cmake:18]: Just curious: does the ${...} here mean that DISTRO_NAME isn't a regular string variable in the context of the `set` command? Why not just `set(DISTRO_NAME "unknown")`?

File: components/scream/ekat/cmake/EkatMpiUtils.cmake [hunk 8]
```diff
@@ -33,23 +34,26 @@ macro (DisableMpiCxxBindings)
 
   if ("${DISTRO_NAME}" STREQUAL "openmpi")
     add_definitions (-DOMPI_SKIP_MPICXX)
-  elseif ("${DISTRO_NAME}" STREQUAL "mpich")
+  elseif ("${DISTRO_NAMEi}" STREQUAL "mpich")
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/EkatMpiUtils.cmake:37]: Is the `i` at the end of `DISTRO_NAME` a typo?

File: components/scream/src/share/grid/remap/abstract_remapper.hpp (hunk 1 above)
Comment (bartgol) [reply components/scream/src/share/grid/remap/abstract_remapper.hpp:7]: and ekat_factory sounds like a cat food processing plant.

File: components/scream/src/physics/shoc/atmosphere_macrophysics.hpp (hunk 2 above)
Comment (bartgol) [reply components/scream/src/physics/shoc/atmosphere_macrophysics.hpp:18]: right, will fix that comment

File: components/scream/ekat/tests/algorithm/lin_interp_test.cpp (hunk 3 above)
Comment (bartgol) [reply components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: We can define a constexpr somewhere, which evaluates to true or false depending on the ifdef. The best place would be ekat_config.h, but that would make parsing it for fortran complicated (we would need two separate ekat_config.h.in and ekat_config.f.in files). Another possible place is `scream_test_utils.hpp`, since FP stuff is mostly needed for testing.

File: components/scream/ekat/cmake/EkatMpiUtils.cmake (hunk 7 above)
Comment (bartgol) [reply components/scream/ekat/cmake/EkatMpiUtils.cmake:18]: DISTRO_NAME contains the name of the variable used to store the result. It is, in a way, a double level of indirection.

CMake does not have a way to _return_ values from functions or macros. When you pass a variable to a function/macro, cmake interpret it as a string. Focusing on macros (function add an extra scope layer, but are otherwise similar), if you have `macro(foo my_var)` and try to do `set(my_var "blah")`, it will work only if, by chance, the variable is called `my_var` also outside the macro. The correct way to write in the variable is to get the content of `my_var` and interpret it as _the name of the output variable_, that is `set (${my_var} "blah")`.

I wish cmake had a simpler way of returning form macros/functions. This can be complicated. Btw, I had it wrong at the beginning (I had what you suggested), but it _happened_ to work, cause at the call site the variable was also called DISTRO_NAME, which worked. When I accidentally changed the var name at the call site to `MPI_DISTRIBUTION_NAME`, everything broke, which helped me track down this cmake subtle issue.

File: components/scream/ekat/cmake/EkatMpiUtils.cmake (hunk 8 above)
Comment (bartgol) [reply components/scream/ekat/cmake/EkatMpiUtils.cmake:37]: Yup, good catch. Can you tell whether I use vim or emacs? =)

File: components/scream/ekat/tests/algorithm/lin_interp_test.cpp (hunk 3 above)
Comment (jeff-cohere) [reply components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: If you think it's worth it, that sounds good.

**Comment by bartgol** [comment]
Ugh, I had to manually merge master (even if it didn't require manual conflict resolution), cause github claimed there were conflicts, so the AT refused to test the branch... :/

File: components/scream/ekat/cmake/EkatMpiUtils.cmake (hunk 7 above)
Comment (jeff-cohere) [reply components/scream/ekat/cmake/EkatMpiUtils.cmake:18]: Gross. Thanks for the explanation!

File: components/scream/src/share/grid/remap/abstract_remapper.hpp (hunk 1 above)
Comment (jeff-cohere) [reply components/scream/src/share/grid/remap/abstract_remapper.hpp:7]: True! Which could be a setting in a bad horror film.

Comment (bartgol) [reply components/scream/src/share/grid/remap/abstract_remapper.hpp:7]: They made millions with Sharknado. I'm sure we could pull off a money make horror film set in an abandoned cat food processing plant. Let's take this offline and start talking movies.

File: components/scream/ekat/cmake/EkatCreateUnitTest.cmake (hunk 6 above)
Comment (bartgol) [reply components/scream/ekat/cmake/EkatCreateUnitTest.cmake:202]: It's mostly a copy+paste of CreateUnitTest from ScreamUtils.cmake (which now calls this one). I find it useful not just for the mpi/omp goodie at the end, but also cause writing

```
EkatCreateUnitTest(myexe "${sources}" INCLUDE_DIRS "${my_include_dirs}" LIBS "foo;bar" LABELS "basic;unit" LINKER_FLAGS "-ld" DEP "setup_test")
```
Is easier to parse for a reader than the sequence

```
add_executable(myexe ${sources})
target_link_libraries(myexe foo bar)
target_include_directories(myexe ${my_include_dirs})
target_link_options(myexe "-ld")
add_test (myexe_test myexe)
set_tests_properties(myexe_test PROPERTIES LABELS "basic;unit" DEPENDS setup_test)
```

The least satisfactory part for me is that it sets test dependencies of all the `A_npN_ompM` tests on a test `B`. But what if B was created with multiple thread/ranks options? There's no way for cmake to figure out the mpi/omp settings for that test, and since `B` is not a valid test name (only `B_npN_ompM` are), it will crap out. I guess in such a scenario, the user is out of luck, and needs to set dependencies manually on all the test, upon return from this function.

File: components/scream/ekat/cmake/EkatMpiUtils.cmake (hunk 7 above)
Comment (bartgol) [reply components/scream/ekat/cmake/EkatMpiUtils.cmake:18]: Gross indeed!

File: components/scream/ekat/tests/algorithm/lin_interp_test.cpp (hunk 3 above)
Comment (bartgol) [reply components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: You mean add it inside test utils?

Comment (jeff-cohere) [reply components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: Yeah, if it's only useful to tests. But as you point out, there are complexities resulting from our multiple languages. Honestly, I could go either way--I was just trying to understand how the macro is used.

Comment (bartgol) [reply components/scream/ekat/tests/algorithm/lin_interp_test.cpp:179]: In Fortran, one can still include `ekat_config.f` and do an ifdef. Maybe create a `logical, parameter :: fp_strict` based on that, in a helper module. But Ok, I think putting it in scream_test_utils.hpp for now is enough.

**Comment by bartgol** [comment]
Turned into a WIP pr, while I figure out what's wrong with the jenkins job on white.

File: components/scream/src/physics/shoc/atmosphere_macrophysics.hpp [hunk 9]
```diff
@@ -2,7 +2,7 @@
 #define SCREAM_SHOC_MACROPHYSICS_HPP
 
 #include "share/atm_process/atmosphere_process.hpp"
-#include "share/scream_parameter_list.hpp"
+#include "ekat/scream_parameter_list.hpp"
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/atmosphere_macrophysics.hpp:5]: This is one of those renames that's being saved for a later PR, right?

File: components/scream/ekat/cmake/Ekat.cmake [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+  string (TOUPPER "${PRECISION}" EKAT_PRECISION)
+
+  if (NOT "${EKAT_PRECISION}" STREQUAL "DOUBLE" AND
+      NOT "${EKAT_PRECISION}" STREQUAL "SINGLE")
+    message (FATAL_ERROR "Error! Unsupported precision for ekat. Valid options are: 'SINGLE', 'DOUBLE'.")
+  endif ()
+
+  # Process EKAT only if not already done
+  if (NOT IS_EKAT_${EKAT_PRECISION}_BUILT)
```
Comment (jeff-cohere) [inline components/scream/ekat/cmake/Ekat.cmake:61]: CMake meta-programming!

File: components/scream/src/physics/shoc/atmosphere_macrophysics.hpp (hunk 9 above)
Comment (bartgol) [reply components/scream/src/physics/shoc/atmosphere_macrophysics.hpp:5]: Right.
